server/src/lib.rs

```diff
@@ -0,0 +1,724 @@
+        println!(
+            "{} network {} is uninstalled.",
+            "[*]".dimmed(),
+            interface.as_str_lossy().yellow()
+        );
+    }
+    Ok(())
+}
+
+pub async fn serve(
+    interface: InterfaceName,
+    conf: &ServerConfig,
+    network: NetworkOpts,
+) -> Result<(), Error> {
+    let config = ConfigFile::from_file(conf.config_path(&interface))?;
+    log::debug!("opening database connection...");
+    let conn = open_database_connection(&interface, conf)?;
+
+    let mut peers = DatabasePeer::list(&conn)?;
+    log::debug!("peers listed...");
+    let peer_configs = peers
+        .iter()
+        .map(|peer| peer.deref().into())
+        .collect::<Vec<PeerConfigBuilder>>();
+
+    log::info!("bringing up interface.");
+    wg::up(
+        &interface,
+        &config.private_key,
+        IpNet::new(config.address, config.network_cidr_prefix)?,
+        Some(config.listen_port),
+        None,
+        network,
+    )?;
+
+    DeviceUpdate::new()
+        .add_peers(&peer_configs)
+        .apply(&interface, network.backend)?;
+
+    log::info!("{} peers added to wireguard interface.", peers.len());
+
+    let candidates: Vec<Endpoint> = get_local_addrs()?
+        .map(|addr| SocketAddr::from((addr, config.listen_port)).into())
+        .collect();
+    let num_candidates = candidates.len();
+    let myself = peers
+        .iter_mut()
+        .find(|peer| peer.ip == config.address)
+        .expect("Couldn't find server peer in peer list.");
+    myself.update(
+        &conn,
+        PeerContents {
+            candidates,
+            ..myself.contents.clone()
+        },
+    )?;
+
+    log::info!(
+        "{} local candidates added to server peer config.",
+        num_candidates
+    );
+
+    let public_key = wireguard_control::Key::from_base64(&config.private_key)?.get_public();
+    let db = Arc::new(Mutex::new(conn));
+    let endpoints = spawn_endpoint_refresher(interface, network);
+    spawn_expired_invite_sweeper(db.clone());
+
+    let context = Context {
+        db,
+        endpoints,
+        interface,
+        public_key,
+        backend: network.backend,
+    };
+
+    log::info!("innernet-server {} starting.", VERSION);
+
+    let listener = get_listener((config.address, config.listen_port).into(), &interface)?;
+
+    let make_svc = hyper::service::make_service_fn(move |socket: &AddrStream| {
+        let remote_addr = socket.remote_addr();
+        let context = context.clone();
+        async move {
+            Ok::<_, http::Error>(hyper::service::service_fn(move |req: Request<Body>| {
+                log::debug!("{} - {} {}", &remote_addr, req.method(), req.uri());
+                hyper_service(req, context.clone(), remote_addr)
+            }))
+        }
+    });
+
+    let server = hyper::Server::from_tcp(listener)?.serve(make_svc);
+
+    server.await?;
+
+    Ok(())
+}
+
+fn spawn_endpoint_refresher(interface: InterfaceName, network: NetworkOpts) -> Endpoints {
```

Comment:
I see this and spawn_expired_invite_sweeper were moved w.r.t. their original place in `main.rs` - could you please restore their original position above `async fn serve(`?

It will make the diff (when viewed with copy detection) smaller for free, thus easier to check.

Reply:
Sure :+1: